### PR TITLE
new style fluff pillow

### PR DESCRIPTION
**Pull request by snopoke**
@czue 

cc @calellowitz 

review by commit

### CODE REVIEW

File: corehq/apps/change_feed/consumer/feed.py [hunk 1]
```diff
@@ -125,9 +125,8 @@ class MultiTopicCheckpointEventHandler(PillowCheckpointEventHandler):
     Event handler that supports checkpoints when subscribing to multiple topics.
     """
 
-    def __init__(self, checkpoint, checkpoint_frequency, change_feed):
-        self.checkpoint = checkpoint
-        self.checkpoint_frequency = checkpoint_frequency
+    def __init__(self, checkpoint, checkpoint_frequency, change_feed, checkpoint_listeners=None):
+        super(MultiTopicCheckpointEventHandler, self).__init__(checkpoint, checkpoint_frequency, checkpoint_listeners)
```
Comment (dimagimon) [inline corehq/apps/change_feed/consumer/feed.py]: E501 line too long (118 > 115 characters)

File: corehq/ex-submodules/pillowtop/processors/interface.py [hunk 2]
```diff
@@ -1,9 +1,64 @@
 from abc import ABCMeta, abstractmethod
+from datetime import datetime, timedelta
+
+from pillowtop.const import CHECKPOINT_FREQUENCY
```
Comment (dimagimon) [inline corehq/ex-submodules/pillowtop/processors/interface.py]: F401 'CHECKPOINT_FREQUENCY' imported but unused

File: corehq/ex-submodules/fluff/pillow.py [hunk 3]
```diff
@@ -137,3 +133,21 @@ def _get_indicator_doc_from_class_and_id(indicator_class, doc_id):
         return indicator_class.get(indicator_id)
     except ResourceNotFound:
         return indicator_class(_id=indicator_id)
+
+
+def get_fluff_pillow(indicator_class, delete_filtered=False, chunk_size=None):
+    processor = FluffPillowProcessor(indicator_class, delete_filtered=delete_filtered, chunk_size=chunk_size)
+
+    name = '{}Pillow'.format(indicator_class.__name__)
+    checkpoint = PillowCheckpoint('fluff.{}.{}'.format(name, get_machine_id()))
```
Comment (snopoke) [inline corehq/ex-submodules/fluff/pillow.py]: checkpoint name remains unchanged from current ones so we shouldn't have any rewindes

File: corehq/ex-submodules/fluff/signals.py [hunk 4]
```diff
@@ -54,17 +54,15 @@ def catch_signal(sender, **kwargs):
...
 
-    from fluff.pillow import FluffPillow
     table_pillow_map = {}
-    pillow_configs = get_all_pillow_configs()
-    for pillow_config in pillow_configs:
-        pillow_class = pillow_config.get_class()
-        if issubclass(pillow_class, FluffPillow):
-            doc = pillow_class.indicator_class()
+    for pillow in get_all_pillow_instances():
+        processor = getattr(pillow, '_processor', None)
```
Comment (snopoke) [inline corehq/ex-submodules/fluff/signals.py]: This get's more elegant later

File: custom/opm/models.py [hunk 5]
```diff
@@ -42,8 +42,6 @@ def user_data(property):
     # Only consider active users
     document_filter = CustomFilter(is_valid_user)
 
-    kafka_topic = topics.META
```
Comment (snopoke) [inline custom/opm/models.py]: this one was incorrect

File: custom/world_vision/models.py [hunk 6]
```diff
@@ -123,7 +122,6 @@ def user_data(property):
     document_class = CommCareUser
     domains = WORLD_VISION_DOMAINS
     group_by = ('domain',)
-    kafka_topic = topics.META
```
Comment (snopoke) [inline custom/world_vision/models.py]: and this one too

File: custom/m4change/models.py [hunk 7]
```diff
@@ -52,7 +52,6 @@ class BaseM4ChangeCaseFluff(fluff.IndicatorDocument):
     document_class = XFormInstance
     document_filter = ORFilter(_get_all_m4change_forms())
     domains = M4CHANGE_DOMAINS
-    kafka_topic = topics.CASE
```
Comment (snopoke) [inline custom/m4change/models.py]: this one also broken

File: corehq/ex-submodules/fluff/pillow.py [hunk 8]
```diff
@@ -135,19 +135,33 @@ def _get_indicator_doc_from_class_and_id(indicator_class, doc_id):
         return indicator_class(_id=indicator_id)
 
 
+class FluffPillow(ConstructedPillow):
+    def __init__(self, indicator_class, processor):
+        self.indicator_class = indicator_class
+        self.kafka_topic = indicator_class().kafka_topic
+        self.domains = processor.domains
+        self.doc_type = processor.doc_type
```
Comment (snopoke) [inline corehq/ex-submodules/fluff/pillow.py]: having these as props on the pillow is useful in a few places (signals and reindexer)

File: corehq/ex-submodules/fluff/management/commands/ptop_reindexer_fluff.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            raise CommandError('Reindexer not configured for topic: {}'.format(pillow.kafka_topic))
+
+        change_provider = CouchDomainDocTypeChangeProvider(
+            couch_db=couch_db,
+            domains=pillow.domains,
+            doc_types=[pillow.doc_type],
+            event_handler=ReindexEventHandler(pillow_name),
+        )
+
+        PillowChangeProviderReindexer(pillow, change_provider).reindex()
```
Comment (snopoke) [inline corehq/ex-submodules/fluff/management/commands/ptop_reindexer_fluff.py]: sweet code reuse

File: corehq/ex-submodules/fluff/management/commands/ptop_fast_reindex_fluff.py [hunk 10]
```diff
@@ -1,76 +0,0 @@
...
-    args = '<domain> <pillow_class>'
-
-    def handle(self, *args, **options):
-        if len(args) != 2:
-            raise CommandError('Usage is ptop_fast_reindex_fluff %s' % self.args)
-
-        self.domain = args[0]
-        self.pillow_class = to_function(args[1])
-
-        super(Command, self).handle(*args, **options)
```
Comment (snopoke) [inline corehq/ex-submodules/fluff/management/commands/ptop_fast_reindex_fluff.py]: :skull:

File: corehq/ex-submodules/pillowtop/listener.py [hunk 11]
```diff
@@ -214,124 +210,6 @@ def change_transport(self, doc_dict):
...
-            if self.use_chunking:
-                self.process_chunk()
-            self.set_checkpoint(change)
-
-    def run(self):
-        self.change_queue = []
-        self.last_processed_time = datetime.utcnow()
-        super(PythonPillow, self).run()
-
-
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/listener.py]: :skull: :skull:

File: corehq/ex-submodules/fluff/management/commands/ptop_reindexer_fluff.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+from django.core.management import CommandError
+from django.core.management.base import BaseCommand
+
+from corehq.apps.change_feed import topics
+from corehq.util.couchdb_management import couch_config
+from corehq.util.pagination import PaginationEventHandler
+from fluff.pillow import get_fluff_pillow_configs
+from pillowtop.reindexer.change_providers.couch import CouchDomainDocTypeChangeProvider
+from pillowtop.reindexer.reindexer import PillowChangeProviderReindexer
+from pillowtop.utils import get_pillow_by_name, get_all_pillow_configs
```
Comment (dimagimon) [inline corehq/ex-submodules/fluff/management/commands/ptop_reindexer_fluff.py]: F401 'get_all_pillow_configs' imported but unused

**Comment by czue** [comment]
Haven't reviewed yet but looks juicy. fyi, looks like the tests are still failling
```
======================================================================
ERROR: test_couverture_fluff (custom.intrahealth.tests.test_fluffs.TestFluffs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/custom/intrahealth/tests/test_fluffs.py", line 101, in test_couverture_fluff
    noinput=True
  File "/vendor/lib/python2.7/site-packages/django/core/management/__init__.py", line 95, in call_command
    raise CommandError("Unknown command: %r" % name)
CommandError: Unknown command: 'ptop_fast_reindex_fluff'
```

File: corehq/ex-submodules/fluff/pillow.py (hunk 3 above)
Comment (czue) [inline corehq/ex-submodules/fluff/pillow.py]: i think this is fine. the other approach if we wanted to get rid of the legacy names would be to do a migration [like this](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/cleanup/migrations/0010_rename_default_change_feed_checkpoint.py#L8)

File: corehq/ex-submodules/pillowtop/processors/interface.py [hunk 13]
```diff
@@ -7,3 +15,53 @@ class PillowProcessor(object):
...
+        pass
+
+
+class ChunkedProcessorCheckpointListener(CheckpointEventListener):
+    def __init__(self, chunked_processor):
+        self.processor = chunked_processor
+
+    def checkpoint_updated(self, updated_to):
+        if self.processor.use_chunking:
+            self.processor.process_chunk()
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/interface.py]: if we do this in the `checkpoint_updated` event does that not mean that we could fail between updating the checkpoint and processing the chunk and then accidentally skip over that chunk when the pillow restarts?

File: corehq/ex-submodules/pillowtop/processors/interface.py [hunk 14]
```diff
@@ -7,3 +15,53 @@ class PillowProcessor(object):
...
+        if self.use_chunking and not is_retry_attempt:
+            self.change_queue.append(change)
+            if self.queue_full or self.wait_expired:
+                self.process_chunk()
+        else:
+            self.process_individual_change(change)
+
+    def process_chunk(self):
+        for change in self.change_queue:
+            self.process_individual_change(change)
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/interface.py]: maybe a dumb question, but if changes still get processed individually what value is chunking providing? i thought it was for bulk loading, but I don't see where that would be happening here.

File: corehq/ex-submodules/fluff/indicators.py [hunk 15]
```diff
@@ -101,6 +100,14 @@ def table(self):
         return self._table
 
     @property
+    def kafka_topic(self):
+        """if set, this will use a kafka feed instead of couch for the pillow"""
```
Comment (czue) [inline corehq/ex-submodules/fluff/indicators.py]: nice

File: custom/m4change/models.py (hunk 7 above)
Comment (czue) [inline custom/m4change/models.py]: :disappointed:

Comment (czue) [inline custom/m4change/models.py]: pretty egregious error here. I guess i'll blame the class name to make myself feel better.

File: corehq/ex-submodules/pillowtop/listener.py (hunk 11 above)
Comment (czue) [inline corehq/ex-submodules/pillowtop/listener.py]: :exclamation: :grey_exclamation: :exclamation:

File: corehq/ex-submodules/pillowtop/processors/interface.py (hunk 14 above)
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/processors/interface.py]: Yea good point. I didn't think to hard about this but I guess the chucking
only helps if we do bulk loads from couch. Getting rid of the chucking
would simplify a lot of things!

**Comment by czue** [comment]
looking good. let me know if you want to remove chunking before this goes in

**Comment by snopoke** [comment]
Yes definitely want to remove the chunking. I'll ping when ready for review.

**Comment by snopoke** [comment]
@czue :white_check_mark:

**Comment by czue** [comment]
:clap: :clap:

**Pull request merged**